engine/rpcserver.go

```diff
@@ -4749,3 +4750,180 @@ func (s *RPCServer) GetTechnicalAnalysis(ctx context.Context, r *gctrpc.GetTechn
 
 	return &gctrpc.GetTechnicalAnalysisResponse{Signals: signals}, nil
 }
+
+// GetMarginRatesHistory returns the margin lending or borrow rates for an exchange, asset, currency along with many customisable options
+func (s *RPCServer) GetMarginRatesHistory(ctx context.Context, r *gctrpc.GetMarginRatesHistoryRequest) (*gctrpc.GetMarginRatesHistoryResponse, error) {
+	if r == nil {
+		return nil, fmt.Errorf("%w GetLendingRatesRequest", common.ErrNilPointer)
+	}
+	exch, err := s.GetExchangeByName(r.Exchange)
+	if err != nil {
+		return nil, err
+	}
+
+	a, err := asset.New(r.Asset)
+	if err != nil {
+		return nil, err
+	}
+
+	err = checkParams(r.Exchange, exch, a, currency.EMPTYPAIR)
+	if err != nil {
+		return nil, err
+	}
+
+	c := currency.NewCode(r.Currency)
+	pairs, err := exch.GetEnabledPairs(a)
+	if err != nil {
+		return nil, err
+	}
+	if !pairs.ContainsCurrency(c) {
+		return nil, fmt.Errorf("%w '%v' in enabled pairs", currency.ErrCurrencyNotFound, r.Currency)
+	}
+
+	start := time.Now().AddDate(0, -1, 0)
+	end := time.Now()
+	if r.StartDate != "" {
+		start, err = time.Parse(common.SimpleTimeFormat, r.StartDate)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if r.EndDate != "" {
+		end, err = time.Parse(common.SimpleTimeFormat, r.EndDate)
+		if err != nil {
+			return nil, err
+		}
+	}
+	err = common.StartEndTimeCheck(start, end)
+	if err != nil {
+		return nil, err
+	}
+
+	request := &margin.RateHistoryRequest{
+		Exchange:           exch.GetName(),
+		Asset:              a,
+		Currency:           c,
+		StartDate:          start,
+		EndDate:            end,
+		GetPredictedRate:   r.GetPredictedRate,
+		GetLendingPayments: r.GetLendingPayments,
+		GetBorrowRates:     r.GetBorrowRates,
+		GetBorrowCosts:     r.GetBorrowCosts,
+		CalculateOffline:   r.CalculateOffline,
+	}
+	if request.CalculateOffline {
+		if r.TakerFeeRate == "" {
+			return nil, fmt.Errorf("%w for offline calculations", common.ErrCannotCalculateOffline)
+		}
+		request.TakeFeeRate, err = decimal.NewFromString(r.TakerFeeRate)
+		if err != nil {
+			return nil, err
+		}
+
+		if request.TakeFeeRate.LessThanOrEqual(decimal.Zero) {
+			return nil, fmt.Errorf("%w for offline calculations", common.ErrCannotCalculateOffline)
+		}
+		if len(r.Rates) == 0 {
+			return nil, fmt.Errorf("%w for offline calculations", common.ErrCannotCalculateOffline)
+		}
+		for i := range r.Rates {
```

Comment:
Can add `request.Rates = make([]margin.Rate, len(r.Rates))` and `request.Rates[i] = offlineRate` at the end of the for loop